flink-core/src/main/java/org/apache/flink/api/common/eventtime/WatermarkStrategies.java

```diff
@@ -0,0 +1,251 @@
+	 *
+	 * <p>You can use this in case you want to specify a {@link TimestampAssigner} via a lambda
+	 * function.
+	 *
+	 * <pre>
+	 * {@code WatermarkStrategy<CustomObject> wmStrategy = WatermarkStrategies
+	 *   .<CustomObject>forMonotonousTimestamps()
+	 *   .withTimestampAssigner((event, timestamp) -> event.getTimestamp())
+	 *   .build();
+	 * }</pre>
+	 */
+	public WatermarkStrategies<T> withTimestampAssigner(SerializableTimestampAssigner<T> timestampAssigner) {
+		checkNotNull(timestampAssigner, "timestampAssigner");
+		this.timestampAssignerSupplier = TimestampAssignerSupplier.of(timestampAssigner);
+		return this;
+	}
+
+	/**
+	 * Build the watermark strategy.
+	 */
+	public WatermarkStrategy<T> build() {
+		WatermarkStrategy<T> strategy = this.baseStrategy;
+
+		if (idleTimeout != null) {
+			strategy = new WithIdlenessStrategy<>(strategy, idleTimeout);
+		}
+
+		if (timestampAssignerSupplier != null) {
+			strategy = new WithTimestampAssigner<>(strategy, timestampAssignerSupplier);
+		}
+
+		return strategy;
+	}
+
+	// ------------------------------------------------------------------------
+	//  builder entry points
+	// ------------------------------------------------------------------------
+
+	/**
+	 * Starts building a watermark strategy for situations with monotonously ascending
+	 * timestamps.
+	 *
+	 * <p>The watermarks are generated periodically and tightly follow the latest
+	 * timestamp in the data. The delay introduced by this strategy is mainly the periodic
+	 * interval in which the watermarks are generated.
+	 *
+	 * @see AscendingTimestampsWatermarks
+	 */
+	public static <T> WatermarkStrategies<T> forMonotonousTimestamps() {
+		return new WatermarkStrategies<T>(AscendingTimestampsWatermarks::new);
+	}
+
+	/**
+	 * Starts building a watermark strategy for situations where records are out of order, but
+	 * you can place an upper bound on how far the events are out of order.
+	 * An out-of-order bound B means that once the an event with timestamp T was encountered, no
+	 * events older than {@code T - B} will follow any more.
+	 *
+	 * <p>The watermarks are generated periodically. The delay introduced by this watermark strategy
+	 * is the periodic interval length, plus the out of orderness bound.
+	 *
+	 * @see BoundedOutOfOrdernessWatermarks
+	 */
+	public static <T> WatermarkStrategies<T> forBoundedOutOfOrderness(Duration maxOutOfOrderness) {
+		return new WatermarkStrategies<>(() -> new BoundedOutOfOrdernessWatermarks<>(maxOutOfOrderness));
+	}
+
+	/**
+	 * Starts building a watermark strategy based on an existing {@code WatermarkStrategy}.
+	 */
+	public static <T> WatermarkStrategies<T> forStrategy(WatermarkStrategy<T> strategy) {
+		return new WatermarkStrategies<>(strategy);
+	}
+
+	/**
+	 * Starts building a watermark strategy based on an existing {@code WatermarkGenerator}.
+	 */
+	public static <T, X extends WatermarkGenerator<T> & Serializable> WatermarkStrategies<T> forGenerator(X generator) {
```

Comment:
Do we still want to add the suppliers here? I think suppliers are even more important for `WatermarkGenerator`.